### PR TITLE
Implementing `KeyProperty` constructor in `ndb`.

### PR DESCRIPTION
**Pull request by dhermes**
This also required adding a small modification to `Property.__repr__` to change how it behaves for `KeyProperty`.

### CODE REVIEW

File: ndb/src/google/cloud/ndb/model.py [hunk 1]
```diff
@@ -2241,9 +2252,240 @@ def __init__(self, *args, **kwargs):
...
+
+        >>> name = "my_value"
+        >>> ndb.KeyProperty(name)
+        KeyProperty(b'my_value')
+        >>> ndb.KeyProperty(SimpleModel)
+        KeyProperty(kind=b'SimpleModel')
+        >>> ndb.KeyProperty(name, SimpleModel)
+        KeyProperty(b'my_value', kind=b'SimpleModel')
+        >>> ndb.KeyProperty(SimpleModel, name)
+        KeyProperty(b'my_value', kind=b'SimpleModel')
```
Comment (dhermes) [inline ndb/src/google/cloud/ndb/model.py]: This `b'my_value'` / `b'SimpleModel'` is showing the same issue we've run into in the unit tests

File: ndb/src/google/cloud/ndb/model.py [hunk 2]
```diff
@@ -2241,9 +2252,240 @@ def __init__(self, *args, **kwargs):
...
+            repeated=repeated,
+            required=required,
+            default=default,
+            choices=choices,
+            validator=validator,
+            verbose_name=verbose_name,
+            write_empty_list=write_empty_list,
+        )
+        if kind is not None:
+            self._kind = kind
```
Comment (dhermes) [inline ndb/src/google/cloud/ndb/model.py]: As in all subtypes that add a data attribute, I'm making this optional even though the original implementation would always set `self._kind` even if `kind` were `None`

Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: Is this useful without a kind? Should it be required?

File: ndb/src/google/cloud/ndb/model.py [hunk 3]
```diff
@@ -2241,9 +2252,240 @@ def __init__(self, *args, **kwargs):
 
 
 class KeyProperty(Property):
-    __slots__ = ()
+    """A property that contains :class:`.Key` values.
 
-    def __init__(self, *args, **kwargs):
+    The constructor for :class:`KeyProperty` allows at most two positional
+    arguments. Any usage of :data:`None` as a positional argument will
```
Comment (chrisrossi) [inline ndb/src/google/cloud/ndb/model.py]: The `None` filtering is weird. Any idea what that's about?

File: ndb/src/google/cloud/ndb/model.py [hunk 4]
```diff
@@ -2241,9 +2252,240 @@ def __init__(self, *args, **kwargs):
...
+            Tuple[str, bool]: Pairs of argument name and a boolean indicating
+            if that argument is a keyword.
+        """
+        yield "name", False
+        yield "kind", True
+        from_inspect = super(KeyProperty, self)._constructor_info()
+        for name, is_keyword in from_inspect:
+            if name in ("args", "name", "kind"):
+                continue
+            yield name, is_keyword
```
Comment (chrisrossi) [inline ndb/src/google/cloud/ndb/model.py]: Looks like that was fun!

File: ndb/tests/unit/test_model.py [hunk 5]
```diff
@@ -1784,9 +1784,142 @@ def test_constructor():
...
+        prop = model.KeyProperty("keyp", kind="Simple", repeated=True)
+        expected = "KeyProperty(b'keyp', kind=b'Simple', repeated=True)"
+        assert repr(prop) == expected
+
+    @staticmethod
+    def test__validate():
+        kind = "Simple"
+        prop = model.KeyProperty("keyp", kind=kind)
+        value = key.Key(kind, 182983)
+        prop._kind = kind  # Ick
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/tests/unit/test_model.py]: Shouldn't we do what datastore does and make `_name` and `_kind` of type `str` here, as well?

File: ndb/src/google/cloud/ndb/model.py (hunk 2 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: @chrisrossi Did you resolve your

> Is this useful without a kind? Should it be required?

question? (I saw it in email but not on GitHub.)

File: ndb/src/google/cloud/ndb/model.py (hunk 3 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: Using

```python
KeyProperty(None)
```

was billed as a way to explicitly set `name=None` in the original docstring. But that is already the default, so I'm not sure why it matters.

File: ndb/src/google/cloud/ndb/model.py (hunk 4 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: Ha yup

File: ndb/tests/unit/test_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply ndb/tests/unit/test_model.py]: :+1: Will do and will make sure to document the problem in `MIGRATION_NOTES`

**Comment by dhermes** [comment]
@chrisrossi PTAL

File: ndb/src/google/cloud/ndb/model.py (hunk 2 above)
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: I decided after looking at it more that it probably was useful and retracted my question.

File: ndb/src/google/cloud/ndb/model.py (hunk 3 above)
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: Probably a backwards compatibility thing, then.

**Pull request merged**